### ISSUE TITLE
get the WAM-IPE_input_parameters into WAM

### ISSUE THREAD

**Issue by ZhuxiaoLi0113** [SWPC-IPE/WAM-IPE#126]
### Observed Behavior
Provide details on what machine you are running on, what configuration (initial conditions and compset scripts)

Indicate the hash ID and branch where you first noticed the problem. 

Describe the incorrect behavior of the WAM-IPE, 
identifying particular source code (if possible) that may be the culprit

### Expected Behavior
When using the same machine and model configuration, indicate what should happen in the source code

If possible, indicate the last known working version on that branch.

Feel free to include additional comments, figures, plots that help describe your issue.

@mentions should be used to draw attention to developers you think can help resolve the issue most effectively.

**Comment by ZhuxiaoLi66** [comment]
Analyzed and revised the WAM physics codes to let the paths of 24HrKpAvg, HemiPow, HemiPowIdx down to the working(ending) physics codes. Also scale the NO in subroutine getno1d by higher 24HrKpAvg level (gt 5).

**Comment by ZhuxiaoLi66** [comment]
Tzu-Wei organized an telecon last Friday, she, Tim and me got update of the parameters relevant current working flow in WAM and IPE from Adam. we also discussed some data interpolation concerns and work flow for future real time runs.

**Comment by ZhuxiaoLi66** [comment]
Now our working branch for this work is raw_high_lat, which has the Adam's latest works on this subject.

**Comment by twfang** [comment]
Possible duplicate of #121

**Comment by ZhuxiaoLi66** [comment]
@akubaryk got it, will let you know after I tested the new branch off raw_high_lat.

**Comment by ZhuxiaoLi66** [comment]
everyone, since the WAM-IPE input parameters don't only include the solar wind drivers, and we already put some comments under the current issue, I suggest to use the current issue plate as a merge of #126 and #121, sounds good to you all?

**Comment by joeschoonover** [comment]
Hash id f29346ad78e21bd1d3b5a01a5e29a8d5cf8f96e2 currently does not compile. Compilation crashes when attempting to compile `idea_ion.f`


```
idea_ion.f(210): error #6404: This name does not have a type, and must have an explicit type.   [HP]
     &   dayno,utsec,F107,KP,HP,HPI,spw_drivers,sda,sza,rlat,zg,grav,      
-----------------------------^
idea_ion.f(232): error #6404: This name does not have a type, and must have an explicit type.   [HP]
     &   dayno,utsec,f107,kp,hp,hpi,spw_drivers,sda,sza,rlat,ht,grav, 
-----------------------------^
idea_ion.f(210): error #6404: This name does not have a type, and must have an explicit type.   [HPI]
     &   dayno,utsec,F107,KP,HP,HPI,spw_drivers,sda,sza,rlat,zg,grav,      
--------------------------------^
idea_ion.f(232): error #6404: This name does not have a type, and must have an explicit type.   [HPI]
     &   dayno,utsec,f107,kp,hp,hpi,spw_drivers,sda,sza,rlat,ht,grav, 
--------------------------------^
idea_ion.f(210): error #6404: This name does not have a type, and must have an explicit type.   [SPW_DRIVERS]
     &   dayno,utsec,F107,KP,HP,HPI,spw_drivers,sda,sza,rlat,zg,grav,      
------------------------------------^
idea_ion.f(232): error #6404: This name does not have a type, and must have an explicit type.   [SPW_DRIVERS]
     &   dayno,utsec,f107,kp,hp,hpi,spw_drivers,sda,sza,rlat,ht,grav, 
------------------------------------^
idea_ion.f(342): error #6362: The data types of the argument(s) are invalid.   [TRIM]
      IF (trim(SPW_DRIVERS)=='swpc_fst') then
```

**Comment by ZhuxiaoLi66** [comment]
Got it, will correct it tomorrow morning.

**Comment by ZhuxiaoLi66** [comment]
@joeschoonover @twfang my WAM updates can be compiled well by command:
 ./NEMS/NEMSAppBuilder rebuild app=wam-ipe
While not the GUI interface method, different compiler?
doing miner revise now, also would like to check Tzu-wei's concern a little bit more, will push to the branch later.

**Comment by joeschoonover** [comment]
Can you push up your latest code so that I can dig into this issue ?

**Comment by ZhuxiaoLi66** [comment]
done the push.

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
Here under this directory, i have collected some plots that would be helpful for validating storm time response of WAM.

https://drive.google.com/drive/folders/1q8Out7QrT8RKMoXhQ-NVgLhfPQEjYLu1?usp=sharing

**Comment by joeschoonover** [comment]
Any work currently being done on raw_high_lat should migrate to raw_high_lat_v2.

Locally, if you are working on raw_high_lat, temporarily save any work in progress
```
git stash
```
Then, check out raw_high_lat_v2
```
git fetch
git checkout origin/raw_high_lat_v2 -b raw_high_lat_v2
```
Then, reapply any of your "work in progress"
```
git stash apply
```

Note that the git stash commands only need to be used if you have uncommitted work in progress.

**Comment by akubaryk** [comment]
Per last week's discussion, split gigawatts (HemiPower and HemiPowerIndex) into NH and SH components. Doesn't seem to be used internally yet in WAM, but defaulted to NH in IPE until we partition the usage per-hemisphere. "hp_wy" and "hpi_wy" have become "nhp_wy" "shp_wy" & "nhpi_wy" "shpi_wy". Interpolation script updated to duplicate GWatt data for each hemisphere, as it is not available per-hemisphere in our historical data files.

**Comment by akubaryk** [comment]
Found a an instance in WAM of HP/HPI in use (NEMS/src/atmos/phys/idea_ion.f), so have defaulted to using "NHP" and "NHPI" for now.

**Comment by ZhuxiaoLi66** [comment]
pushed the new Joule Heating factor in idea_ion.f to origin/raw_high_lat_2.

**Comment by ZhuxiaoLi66** [comment]
raw_high_lat_v2

**Comment by ZhuxiaoLi66** [comment]
seems the push didn't succeed...

**Comment by akubaryk** [comment]
@ZhuxiaoLi66 you will need to `git pull` before you `git push` it back because I was committing something at the same time -- fixing a small collection of errors in interpolate_input_parameters.py. My fault for not testing it properly yesterday.

**Comment by ZhuxiaoLi66** [comment]
@akubaryk thanks, I tried repeat git pull and git push several times, after git push, it said 'everything updated', while it is not. maybe before this I did 'git stash' and 'git stash apply' to merge my new stuff in local raw_high_lat to  raw_high_lat_v2 locally, the merge seems worked well by checking the code.

**Comment by akubaryk** [comment]
Can you share the directory of your WAM-IPE repository, @ZhuxiaoLi66? Might be able to find something if I poke around. Can discuss on Slack if it's easier.

**Comment by ZhuxiaoLi66** [comment]
/scratch3/NCEPDEV/swpc/save/Zhuxiao.Li/git/WAM-IPE_JH3_2
Thanks, Adam.

**Comment by ZhuxiaoLi66** [comment]
@akubaryk @twfang  the phys codes need more revise to to adapt adam's update.

**Comment by akubaryk** [comment]
Just pushed the updated idea_ion.f to raw_high_lat_v2, not sure if the other untracked files need to be added as well?

**Comment by ZhuxiaoLi66** [comment]
sorry, it seems that my idea_phys.f not get updated. @akubaryk's code should be fine.

**Comment by twfang** [comment]
@ZhuxiaoLi66, I notice that there are many print statements associated with the drivers. Do you mind to clean them up and update these files again? Thanks!

**Comment by twfang** [comment]
A 15-day run of WAM from 2013/03/06 to 2013/03/20 is submitted. Once the run is finished, I will need @ZhuxiaoLi66 to visualize the outputs and @timfullerrowell to evaluate the WAM storm-time responses.

**Comment by ZhuxiaoLi66** [comment]
@akubaryk thanks , I need more practice.  just now I succeeded to do another push. Just delete some print out commands after the driver reading in validation.

**Comment by ZhuxiaoLi66** [comment]
@twfang maybe I can show some physics with my current NCL codes.

**Comment by ZhuxiaoLi66** [comment]
![t_2d_wam_lev135mar14-19_6days_varying_drivers](https://user-images.githubusercontent.com/22546571/36861344-82c5856e-1d7b-11e8-9abe-19d6cdc265b3.png)

**Comment by ZhuxiaoLi66** [comment]
the above plot is the WAM neutral temperature at about 250km during 2013Mar14-19 storm period with the varying drivers in. The temp. responded to the drivers!  The relationship between the time varying drivers and the Temp. need to look into further.

**Comment by ZhuxiaoLi66** [comment]
![t_2d_wam_lev145mar14-19_6days_varying_drivers](https://user-images.githubusercontent.com/22546571/36861655-6cc6378a-1d7c-11e8-8998-17ec4c457f8f.png)
same plot but at about 360km height level.

**Comment by twfang** [comment]
The temperature enhancement seems to be too small. @ZhuxiaoLi66, can you calculate the mean molecular mass and plot them in the same format?

**Comment by ZhuxiaoLi66** [comment]
@twfang I agree about the amplitude of the Temp. enhancement.

**Comment by ZhuxiaoLi66** [comment]
![t_2d_wam_135lev_mar15-20_6days_varying_drivers](https://user-images.githubusercontent.com/22546571/37171399-a91b4818-2305-11e8-9c09-e2af4a07802b.png)

**Comment by ZhuxiaoLi66** [comment]
![v_2d_wam_135lev_mar15-20_6days_varying_drivers](https://user-images.githubusercontent.com/22546571/37171533-1fccea3e-2306-11e8-8124-68b26684e1c1.png)

@twfang @timfullerrowell the plots are meridional wind at lev135 (about 250km).

**Comment by ZhuxiaoLi66** [comment]
compiled the new code for get the solar wind parameters into WAM run (subr prep_weimer in efield.f), submitted a validation run.

**Comment by ZhuxiaoLi66** [comment]
The new 'solar wind' code enhance the neutral temp. response to the storm.

**Comment by ZhuxiaoLi66** [comment]
![t_v_lev135_2d_polar_solar_wind_013031800](https://user-images.githubusercontent.com/22546571/37359607-f887bc9a-26e5-11e8-9ae0-a2e44a451a36.png)

**Comment by ZhuxiaoLi66** [comment]
![t_v_lev135_2d_polar_2013031800_without_sw](https://user-images.githubusercontent.com/22546571/37359624-05f83ac6-26e6-11e8-801a-587e98cbcea6.png)

**Comment by twfang** [comment]
The plots look great @ZhuxiaoLi2 ! What is the latitude at the outer most circle? Could you please also include the same plot for quiet time? Thank you!

**Comment by ZhuxiaoLi66** [comment]
@twfang  it is north hemisphere, the latitude of outer most circle is 60N.

**Comment by ZhuxiaoLi66** [comment]
@twfang Here is the quite time plot from your run.
 
![t_v_lev135_2d_polar_2013031100_without_sw_0306](https://user-images.githubusercontent.com/22546571/37368191-684f01a2-26fd-11e8-835e-5072f49fff9b.png)

**Comment by ZhuxiaoLi66** [comment]
@timfullerrowell @twfang After I added more printout commands in idea_phys.f and recompile WAM,
All the kp, f10 and solar wind parameters are varying now by kdt_interval (number of time step). so the WAM code for this should be fine.

**Comment by timfullerrowell** [comment]
This comment applies to the real-time run, rather than for historical intervals. 

Here are the conversions from Kp forecast to auroral GW and power index, and solar wind parameters to populate the 7-day real-time data file (current day + 3 days before + 3 days after):

On the SW and GW forecasts, here are the conversions from the Kp and F10.7 forecast to the various SW and GW parameters.

1. Ovation forecast values, use the forecasted values for the period they are available.  Beyond that use:

HemiPow (GW) = 1.29 + 15.60 * Kp - 4.93 * Kp_squared + 0.64 * Kp_cubed

Same in north and south as before for now.

2. SW_Bz = -0.85 * Kp_squared - 0.081* Kp + 0.434 + 0.0079 * F10.7(daily) + 0.0022 * Kp * F10.7(daily)

3. SW_By = 0.0

4. SW_Bt = SW_Bz

5. If (Bz .ge. 0.0) SW_Angle = 0.0
    If (Bz .lt. 0.0) SW_Angle = 180.

6. SW_Velocity = 317.0 + 55.84 * Kp - 2.71 * Kp_squared

7. SW_Density = 5.0

8. HemiPowIdx is based on the HemiPow from this table.

Index  between GW 
1  between 0.0 and 2.5
2  between 2.5 and 3.94
3  between 3.94 and 6.22
4  same pattern.... 9.82
5  ............15.49
6  ............24.44
7  ............38.56
8  ............60.85
9  ............96.0
10 ......greater than 96.

For all of these, so we don't have a sudden transition from observed to predicted, I suggested we linearly transition from the last observed value to the predicted value over a 3 hour period. Have to think about the SW_Angle here, when By is still non-zero, tan SW_Angle = By/Bz.

We will also need to fill the values in this file for the last day. For instance, at 6UT  the 3-day Kp forecast will have values up to 6UT on the last day in the file. The values from 6UT to 24UT on this last day will have to be filled. I suggest we decay from the latest forecasted value to a climatological value for each of the parameters, with something like a 12 hour time constant (TBD). I will come up with the climatological values.

**Comment by ZhuxiaoLi66** [comment]
@akubaryk I open an new pull request, please double check my revise in ./scripts/compsets/exglobal/exglobal_fcst_nems.sh
I added one line 'export F107_KP_DATA_SIZE=$((LEN_F107-5))' to avoid WAM to read only one hour input from the parameter file. Thank you.

**Comment by ZhuxiaoLi66** [comment]
Now we get the effective output (parameters varying during the run) for the updated physics code. 
Both for noSW(withought solar wind, only f107 and kp related parameters) and SW (include Solar Wind parameters in prep_weimer). the neutral T&horizontal wind field has been checked. The following are the plots for corresponding 2013 St. Patrick Day storm runs (20130301-20130320). 
The T&wind field of SW run seems more reasonable. will check the T&wind field further and also the compositions. 
![t_wind_polar_wam_lev135_mar17_sw_0301_vary_1min](https://user-images.githubusercontent.com/22546571/37924527-a9398164-3121-11e8-907d-a3556288bc2c.png)
![t_wind_polar_wam_lev135_mar17_no_sw_0301_new2](https://user-images.githubusercontent.com/22546571/37924537-afea128a-3121-11e8-9cab-16d017becb0d.png)

**Comment by ZhuxiaoLi66** [comment]
![t_wind_polar_wam_lev135_mar18_sw_0301_vary_1min](https://user-images.githubusercontent.com/22546571/37925301-b5459c5c-3123-11e8-943f-b58411ac5f20.png)
![t_wind_polar_wam_lev135_mar18_no_sw_0301_new2](https://user-images.githubusercontent.com/22546571/37925309-badf172e-3123-11e8-9ed0-891db5b4545d.png)

**Comment by twfang** [comment]
60hr crash issue in standalone WAM would be resolved after reversing the change in exglobal script that Adam committed. 
https://github.com/SWPC-IPE/WAM-IPE/commit/4c398e857ed6e5852318634f6a159e35204f91de

I confirm that the WAM 5 day run with daily restart finished successfully after the change is made.

**Comment by akubaryk** [comment]
Reversing that commit uses incorrect input parameters for the runs.

**Comment by twfang** [comment]
@akubaryk So the WAM run would finish but with wrong input parameters? But with the commit, the model will stop right at 60hr. Any suggestions?

**Comment by twfang** [comment]
I think we should reverse the commit, merge the raw_high_lat_v3 branch, and then tackle this commit along with the restart issue.

**Comment by ZhuxiaoLi66** [comment]
@akubaryk I agree. we need to figure out the exact reason together.
